pkg/operator/customroute/custom_route_controller.go

```diff
@@ -0,0 +1,283 @@
+	componentRouteName string,
+	destSecretNamespace string,
+	destSecretName string,
+	targetRoute *routev1.Route,
+	consumingUsers []configv1.ConsumingUser,
+	ingressInformer configinformers.IngressInformer,
+	ingressClient configsetterv1.IngressInterface,
+	routeInformer routeinformer.RouteInformer,
+	routeClient routeclient.RouteInterface,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder,
+	resourceSyncer resourcesynccontroller.ResourceSyncer,
+) factory.Controller {
+	controller := &customRouteController{
+		destSecret: types.NamespacedName{
+			Namespace: destSecretNamespace,
+			Name:      destSecretName,
+		},
+		componentRoute: types.NamespacedName{
+			Namespace: componentRouteNamespace,
+			Name:      componentRouteName,
+		},
+		targetRoute:    targetRoute.DeepCopy(),
+		consumingUsers: consumingUsers,
+		ingressLister:  ingressInformer.Lister(),
+		ingressClient:  ingressClient,
+		routeLister:    routeInformer.Lister(),
+		routeClient:    routeClient,
+		secretLister:   kubeInformersForNamespaces.SecretLister(),
+		operatorClient: operatorClient,
+		resourceSyncer: resourceSyncer,
+	}
+
+	return factory.New().
+		WithInformers(
+			ingressInformer.Informer(),
+			routeInformer.Informer(),
+			kubeInformersForNamespaces.InformersFor("openshift-config").Core().V1().Secrets().Informer(),
+			kubeInformersForNamespaces.InformersFor(destSecretNamespace).Core().V1().Secrets().Informer(),
+		).
+		WithSyncDegradedOnError(operatorClient).
+		WithSync(controller.sync).
+		ResyncEvery(time.Minute).
+		ToController("CustomRouteController", eventRecorder.WithComponentSuffix("custom-route-controller"))
+}
+
+func (c *customRouteController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	ingressConfig, err := c.ingressLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	ingressConfigCopy := ingressConfig.DeepCopy()
+
+	// configure the expected route
+	expectedRoute, secretName, errors := c.getOAuthRouteAndSecretName(ingressConfigCopy)
+	if errors != nil {
+		// log if there is an issue updating the ingressConfig resource
+		if updateIngressConfigErr := c.updateIngressConfigStatus(ctx, ingressConfigCopy, errors); updateIngressConfigErr != nil {
+			klog.Infof("Error updating ingress with custom route status: %v", err)
+		}
+		return fmt.Errorf("custom route configuration failed verification: %v", errors)
+	}
+
+	// create or modify the existing route
+	if _, _, err = resourceapply.ApplyRoute(ctx, c.routeClient, syncCtx.Recorder(), expectedRoute); err != nil {
+		return err
+	}
+
+	// update ingressConfig status
+	if err = c.updateIngressConfigStatus(ctx, ingressConfigCopy, nil); err != nil {
+		return err
+	}
+
+	// sync the secret
+	return c.syncSecret(secretName)
+}
+
+func (c *customRouteController) getOAuthRouteAndSecretName(ingressConfig *configv1.Ingress) (*routev1.Route, string, []error) {
+	route := c.targetRoute
```

Comment:
should we use bindata for this instead?

Reply:
Actually no, we can't do it here. 

If we want this code to be reusable (by the Console Team), we need to obtain the route from the Operator that will be using this code. In our case, this will be `openshift-authentication/oauth-openshift`, but in the Console's case it will be something different. 

By the way, the library-go doesn't seem to ship bindata at all. But that's a side note.

Reply:
> If we want this code to be reusable (by the Console Team), we need to obtain the route from the Operator that will be using this code.

...which could be done by using bindata, it's very likely that the `route` here will actually be loaded from bindata while starting this operator.

See https://github.com/openshift/console-operator/blob/c92d52d4a83afb77e6b586d37f1f0dfd6d891df2/pkg/console/subresource/route/route.go#L143-L152, they're using bindata already. I was helping them a little to move to it.

> By the way, the library-go doesn't seem to ship bindata at all. But that's a side note.

That's not entirely true, there are subpackages that ship their own bindata, like `pkg/operator/apiserver/audit/bindata` for example. Also, there are other controllers that are capable of handling bindata, like https://github.com/openshift/library-go/blob/14aa5b749d8ddbae509ea97c988228934dbbd3db/pkg/operator/staticresourcecontroller/static_resource_controller.go#L68

Reply:
I'm still not entirely convinced. What if one of the operators (either Console or CAO) moves to Go Embed and the other doesn't? 

I'm still in favor of pushing the responsibility of obtaining the Route to the caller. The only improvement that I can think of here is to make the Route parameter in the constructor a function pointer. Then, the caller could put there whatever he wants. Let me know if you'd like me to implement this change.

Reply:
I wouldn't mind forcing Go embed, but for the time being the library-go is using bindata for this purpose on other places so I assume the move will be unite for these at certain point.

The reason I don't like this is that it's unnecessary state that's being kept in a somewhat global variable and is therefore prone to side effects. I've done my part in functional programming and so I consider this ugly.

That being said, I can live with it.

Reply:
As agreed to put the route into the bindata. If the Console Operator requires something else - we'll adapt.

Reply:
Is this still the same?

Reply:
Nope, it's not, see https://github.com/openshift/library-go/pull/1138#issuecomment-910024433